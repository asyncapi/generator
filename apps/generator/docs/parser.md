---
title: "Parser"
weight: 60
---

## Parser

The **AsyncAPI Parser** is a powerful library for parsing and validating [AsyncAPI documents](asyncapi-document) in both Node.js and browser environments. It supports YAML and JSON inputs and ensures correctness through schema validation.

### âœ… Supported schema formats:

- AsyncAPI (v2 and v3)
- OpenAPI
- JSON Schema
- Avro
- RAML

These schemas are automatically transformed into JSON Schema format internally, simplifying access and manipulation.

---

## Key Features

- **Parsing** of AsyncAPI YAML/JSON files.
- **Validation** against AsyncAPI and referenced schemas.
- **Dereferencing** of `$ref` values into inlined content.
- **Helper functions** to access channels, messages, schemas, and more.

---

## Example: AsyncAPI v3 Document

Below is a minimal v3.0.0-compliant AsyncAPI document using `channels`, `operations`, and component references:

```yaml
asyncapi: 3.0.0
info:
  title: Comments Service
  version: 1.0.0
  description: Handles all comment-related events.
servers:
  dev:
    host: test.mosquitto.org
    protocol: mqtt
channels:
  sendCommentLiked:
    address: comment/liked
    messages:
      commentLiked:
        $ref: '#/components/messages/commentLiked'
  sendCommentUnliked:
    address: comment/unliked
    messages:
      commentUnliked:
        $ref: '#/components/messages/commentUnliked'
operations:
  sendCommentLiked:
    action: send
    channel:
      $ref: '#/channels/sendCommentLiked'
  sendCommentUnliked:
    action: send
    channel:
      $ref: '#/channels/sendCommentUnliked'
components:
  messages:
    commentLiked:
      payload:
        $ref: '#/components/schemas/commentReaction'
    commentUnliked:
      payload:
        $ref: '#/components/schemas/commentReaction'
  schemas:
    commentReaction:
      type: object
      properties:
        commentId:
          type: string
