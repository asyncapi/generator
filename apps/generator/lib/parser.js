const fs = require('fs');
const { convertToOldAPI } = require('@asyncapi/parser');
const { ConvertDocumentParserAPIVersion, NewParser } = require('@smoya/multi-parser');

const parser = module.exports;

/**
 * Convert the template defined value `apiVersion: 'v1'` to only contain the numeric value `1`.
 */
parser.sanitizeTemplateApiVersion = (apiVersion) => {
  if (!apiVersion) return;
  if (apiVersion && apiVersion.length > 1) {
    return Number(apiVersion.substring(1));
  }
  return Number(apiVersion);
};

parser.parse = async (asyncapi, oldOptions, generator) => {
  let apiVersion = this.sanitizeTemplateApiVersion(generator.templateConfig.apiVersion);
  // Defaulting to apiVersion v1 to convert it to the Parser-API v1 afterwards.
  if (!this.usesNewAPI(generator.templateConfig)) {
    apiVersion = 1;
  }
  const options = convertOldOptionsToNew(oldOptions, generator);
  const parser = NewParser(apiVersion, {parserOptions: options, includeSchemaParsers: true});
  const { document, diagnostics } = await parser.parse(asyncapi, options);
  if (!document) {
    return {document, diagnostics};
  }
  const correctDocument = this.getProperApiDocument(document, generator.templateConfig);
  return {document: correctDocument, diagnostics};
};

/**
 * If the template expect one of the Parser-API versions, it must be above 0
 */
parser.usesNewAPI = (templateConfig = {}) => {
  return this.sanitizeTemplateApiVersion(templateConfig.apiVersion) > 0;
};

/**
 * Based on the current parsed AsyncAPI document, convert it to expected API version from the template.
 */
parser.getProperApiDocument = (asyncapiDocument, templateConfig = {}) => {
  const apiVersion = this.sanitizeTemplateApiVersion(templateConfig.apiVersion);
  if (apiVersion === undefined) {
    // Convert to old API from JS Parser v1
    return convertToOldAPI(asyncapiDocument);
  }
  return ConvertDocumentParserAPIVersion(asyncapiDocument, apiVersion);
};

// The new options for the v2 Parser are different from those for the v1 version, but in order not to release Generator v2, we are converting the old options of Parser to the new ones.
function convertOldOptionsToNew(oldOptions, generator) {
  if (!oldOptions) return;
  const newOptions = {};

  if (typeof oldOptions.path === 'string') {
    newOptions.source = oldOptions.path;
  }
  if (typeof oldOptions.applyTraits === 'boolean') {
    newOptions.applyTraits = oldOptions.applyTraits;
  }

  const resolvers = [];
  if (generator && generator.mapBaseUrlToFolder && generator.mapBaseUrlToFolder.url) {
    resolvers.push(...getMapBaseUrlToFolderResolvers(generator.mapBaseUrlToFolder));
  }
  if (oldOptions.resolve) {
    resolvers.push(...convertOldResolvers(oldOptions.resolve));
  }

  if (resolvers.length) {
    newOptions.__unstable = {};
    newOptions.__unstable.resolver = {
      resolvers,
    };
  }

  return newOptions;
}

/**
 * Creates a custom resolver that maps urlToFolder.url to urlToFolder.folder
 * Building your custom resolver is explained here: https://apitools.dev/json-schema-ref-parser/docs/plugins/resolvers.html
 *
 * @private
 * @param  {object} urlToFolder to resolve url e.g. https://schema.example.com/crm/ to a folder e.g. ./test/docs/.
 * @return {{read(*, *, *): Promise<unknown>, canRead(*): boolean, order: number}}
 */
function getMapBaseUrlToFolderResolvers({ url: baseUrl, folder: baseDir }) {
  const resolver = {
    order: 1,
    canRead: true,
    read(uri) {
      return new Promise(((resolve, reject) => {
        const path = uri.toString();
        const localpath = path.replace(baseUrl, baseDir);
        try {
          fs.readFile(localpath, (err, data) => {
            if (err) {
              reject(`Error opening file "${localpath}"`);
            } else {
              resolve(data.toString());
            }
          });
        } catch (err) {
          reject(`Error opening file "${localpath}"`);
        }
      }));
    }
  };

  return [
    { schema: 'http', ...resolver, },
    { schema: 'https', ...resolver, },
  ];
};

function convertOldResolvers(resolvers = {}) { // NOSONAR
  if (Object.keys(resolvers).length === 0) return [];

  return Object.entries(resolvers).map(([protocol, resolver]) => {
    return {
      schema: protocol,
      order: resolver.order || 1,
      canRead: (uri) => canReadFn(uri, resolver.canRead),
      read: (uri) => {
        return resolver.read({ url: uri.valueOf(), extension: uri.suffix() });
      }
    };
  });
}

function canReadFn(uri, canRead) {
  const value = uri.valueOf();
  if (typeof canRead === 'boolean') return canRead;
  if (typeof canRead === 'string') return canRead === value;
  if (Array.isArray(canRead)) return canRead.includes(value);
  if (canRead instanceof RegExp) return canRead.test(value);
  if (typeof canRead === 'function') {
    return canRead({ url: value, extension: uri.suffix() });
  }
  return false;
}
