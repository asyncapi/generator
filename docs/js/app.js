
    const schema = {
  "asyncapi": "3.0.0",
  "info": {
    "title": "Temperature Service",
    "version": "1.0.0",
    "description": "Service that emits temperature changes from a bedroom sensor."
  },
  "servers": {
    "production": {
      "host": "broker.example.com",
      "protocol": "mqtt"
    }
  },
  "channels": {
    "temperatureChanged": {
      "address": "temperature/changed",
      "messages": {
        "temperatureChange": {
          "description": "Message sent when the temperature in the bedroom changes.",
          "payload": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
              "value": {
                "type": "number",
                "x-parser-schema-id": "<anonymous-schema-1>"
              },
              "unit": {
                "type": "string",
                "x-parser-schema-id": "<anonymous-schema-2>"
              }
            },
            "x-parser-schema-id": "Temperature"
          },
          "x-parser-unique-object-id": "temperatureChange",
          "x-parser-message-name": "temperatureChange"
        }
      },
      "x-parser-unique-object-id": "temperatureChanged"
    }
  },
  "operations": {
    "sendTemperatureChanged": {
      "action": "send",
      "summary": "Temperature changes are pushed to the broker",
      "channel": "$ref:$.channels.temperatureChanged",
      "x-parser-unique-object-id": "sendTemperatureChanged"
    }
  },
  "components": {
    "schemas": {
      "Temperature": "$ref:$.channels.temperatureChanged.messages.temperatureChange.payload"
    }
  },
  "x-parser-spec-parsed": true,
  "x-parser-api-version": 3,
  "x-parser-spec-stringified": true
};
    const config = {"show":{"sidebar":true},"sidebar":{"showOperations":"byDefault"}};
    const appRoot = document.getElementById('root');
    AsyncApiStandalone.render(
        { schema, config, }, appRoot
    );
  