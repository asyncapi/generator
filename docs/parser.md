---
title: "Parser"
weight: 90
---

>This document is under construction.