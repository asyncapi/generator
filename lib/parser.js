const fs = require('fs');

const { Parser, convertToOldAPI } = require('@asyncapi/parser/cjs');
const { OpenAPISchemaParser } = require('@asyncapi/openapi-schema-parser');
const { AvroSchemaParser } = require('@asyncapi/avro-schema-parser');
const { RamlDTSchemaParser } = require('@asyncapi/raml-dt-schema-parser');
const { ProtoBuffSchemaParser } = require('@asyncapi/protobuf-schema-parser');

const parser = module.exports;

const defaultParser = new Parser({
  schemaParsers: [
    OpenAPISchemaParser(),
    AvroSchemaParser(),
    RamlDTSchemaParser(),
    ProtoBuffSchemaParser(),
  ],
});

parser.parse = (asyncapi, oldOptions, generator) => {
  const options = convertOldOptionsToNew(oldOptions, generator);
  return defaultParser.parse(asyncapi, options);
};

parser.usesNewAPI = (templateConfig = {}) => {
  return templateConfig.apiVersion === 'v1';
};

parser.getProperApiDocument = (asyncapiDocument, templateConfig) => {
  return parser.usesNewAPI(templateConfig) ? asyncapiDocument : convertToOldAPI(asyncapiDocument);
};

// The new options for the v2 Parser are different from those for the v1 version, but in order not to release Generator v2, we are converting the old options of Parser to the new ones.
function convertOldOptionsToNew(oldOptions, generator) {
  if (!oldOptions) return;
  const newOptions = {};

  if (typeof oldOptions.path === 'string') {
    newOptions.source = oldOptions.path;
  }
  if (typeof oldOptions.applyTraits === 'boolean') {
    newOptions.applyTraits = oldOptions.applyTraits;
  }

  const resolvers = [];
  if (generator && generator.mapBaseUrlToFolder && generator.mapBaseUrlToFolder.url) {
    resolvers.push(...getMapBaseUrlToFolderResolvers(generator.mapBaseUrlToFolder));
  }
  if (oldOptions.resolve) {
    resolvers.push(...convertOldResolvers(oldOptions.resolve));
  }

  if (resolvers.length) {
    newOptions.__unstable = {};
    newOptions.__unstable.resolver = {
      resolvers,
    };
  }

  return newOptions;
}

/**
 * Creates a custom resolver that maps urlToFolder.url to urlToFolder.folder
 * Building your custom resolver is explained here: https://apitools.dev/json-schema-ref-parser/docs/plugins/resolvers.html
 *
 * @private
 * @param  {object} urlToFolder to resolve url e.g. https://schema.example.com/crm/ to a folder e.g. ./test/docs/.
 * @return {{read(*, *, *): Promise<unknown>, canRead(*): boolean, order: number}}
 */
function getMapBaseUrlToFolderResolvers({ url: baseUrl, folder: baseDir }) {
  const resolver = {
    order: 1,
    canRead: true,
    read(uri) {
      return new Promise(((resolve, reject) => {
        const path = uri.toString();
        const localpath = path.replace(baseUrl, baseDir);
        try {
          fs.readFile(localpath, (err, data) => {
            if (err) {
              reject(`Error opening file "${localpath}"`);
            } else {
              resolve(data.toString());
            }
          });
        } catch (err) {
          reject(`Error opening file "${localpath}"`);
        }
      }));
    }
  };

  return [
    { schema: 'http', ...resolver, },
    { schema: 'https', ...resolver, },
  ];
};

function convertOldResolvers(resolvers = {}) { // NOSONAR
  if (Object.keys(resolvers).length === 0) return [];

  return Object.entries(resolvers).map(([protocol, resolver]) => {
    return {
      schema: protocol,
      order: resolver.order || 1,
      canRead: (uri) => canReadFn(uri, resolver.canRead),
      read: (uri) => {
        return resolver.read({ url: uri.valueOf(), extension: uri.suffix() });
      }
    };
  });
}

function canReadFn(uri, canRead) {
  const value = uri.valueOf();
  if (typeof canRead === 'boolean') return canRead;
  if (typeof canRead === 'string') return canRead === value;
  if (Array.isArray(canRead)) return canRead.includes(value);
  if (canRead instanceof RegExp) return canRead.test(value);
  if (typeof canRead === 'function') {
    return canRead({ url: value, extension: uri.suffix() });
  }
  return false;
}
