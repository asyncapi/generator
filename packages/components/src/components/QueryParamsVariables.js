import { Text } from '@asyncapi/generator-react-sdk';
import { toCamelCase } from '@asyncapi/generator-helpers';

/**
 * @typedef {'python' | 'java' | 'javascript'} SupportedLanguage
 * Supported programming languages for query parameter generation.
 */

/**
 * @typedef {Object} QueryParamCodeBlock
 * @property {{ text: string, indent?: number, newLines?: number }} variableDefinition - Code block for variable initialization.
 * @property {{ text: string, indent?: number, newLines?: number }} ifCondition - Conditional statement block.
 * @property {{ text: string, indent?: number, newLines?: number }} assignment - Code block assigning query param.
 * @property {{ text: string, indent?: number, newLines?: number } | null} [closing] - Optional closing block (for braces, etc.).
 */

/**
 * Language and framework specific logic for generating query parameter code.
 * Each entry returns a {@link QueryParamCodeBlock}.
 *
 * @type {Record<SupportedLanguage, Record<string, function>|function>}
 */
const queryParamLogicConfig = {
  python: (param) => {
    const paramName = param[0];
    return {
      variableDefinition: {
        text: `${paramName} = ${paramName} or os.getenv("${paramName.toUpperCase()}")`,
        indent: 8,
      },
      ifCondition: {
        text: `if ${paramName} is not None:`,
        indent: 8,
      },
      assignment: {
        text: `params["${paramName}"] = ${paramName}`,
        indent: 10,
      },
      closing: null,
    };
  },
  java: {
    quarkus: (param) => {
      const paramName = toCamelCase(param[0]);
      return {
        variableDefinition: {
          text: `this.${paramName} = (${paramName} != null && !${paramName}.isEmpty()) ? ${paramName} : System.getenv("${paramName.toUpperCase()}");`,
          indent: 6,
        },
        ifCondition: {
          text: `if (this.${paramName} != null){`,
          indent: 6,
        },
        assignment: {
          text: `params.put("${paramName}", this.${paramName});`,
          indent: 8,
        },
        closing: {
          text: '}',
          indent: 6,
          newLines: 1,
        },
      };
    },
  },
  javascript: (param) => {
    const paramName = param[0];
    return {
      variableDefinition: {
        text: `const ${paramName} = ${paramName} || process.env.${paramName.toUpperCase()};`,
        indent: 8,
      },
      ifCondition: {
        text: `if (${paramName}) {`,
        indent: 8,
      },
      assignment: {
        text: `params["${paramName}"] = ${paramName};`,
        indent: 10,
      },
      closing: {
        text: '}',
        indent: 8,
        newLines: 1,
      },
    };
  },
};

/**
 * Resolve the appropriate query parameter configuration function based on language and framework.
 *
 * @param {SupportedLanguage} language - The target programming language.
 * @param {string} [framework=''] - Optional framework (e.g., 'quarkus' for Java).
 * @returns {function} The configuration function for generating query parameter code.
 */
function resolveQueryParamLogic(language, framework = '') {
  const config = queryParamLogicConfig[language];
  if (typeof config === 'function') {
    return config;
  }
  if (framework && config[framework]) {
    return config[framework];
  }
}

/**
 * Component for rendering query parameter variables code.
 *
 * @param {Object} props - Component props.
 * @param {SupportedLanguage} props.language - The target programming language.
 * @param {string} [props.framework=''] - Optional framework for the language.
 * @param {string[][]} props.queryParams - Array of query parameters, each represented as [paramName, paramType?].
 */
export function QueryParamsVariables({ language, framework = '', queryParams }) {
  if (!queryParams || !Array.isArray(queryParams)) {
    return null;
  }

  const generateParamCode = resolveQueryParamLogic(language, framework);

  return queryParams.map((param) => {
    const { variableDefinition, ifCondition, assignment, closing } = generateParamCode(param);

    return (
      <>
        <Text indent={variableDefinition.indent} newLines={variableDefinition.newLines}>
          {variableDefinition.text}
        </Text>
        <Text indent={ifCondition.indent} newLines={ifCondition.newLines}>
          {ifCondition.text}
        </Text>
        <Text indent={assignment.indent} newLines={assignment.newLines}>
          {assignment.text}
        </Text>
        {closing && (
          <Text indent={closing.indent} newLines={closing.newLines}>
            {closing.text}
          </Text>
        )}
      </>
    );
  });
}