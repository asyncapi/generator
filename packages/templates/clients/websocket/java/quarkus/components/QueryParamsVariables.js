import { toCamelCase } from '@asyncapi/generator-helpers';
import { Text } from '@asyncapi/generator-react-sdk';

export function QueryParamsVariables({ queryParams }) {
  if (!queryParams || !Array.isArray(queryParams)) { 
    return null;
  }

  return queryParams.map((param, index) => {
    const paramName = toCamelCase(param[0]);
    const variableDefinition = `this.${paramName} = (${paramName} != null && !${paramName}.isEmpty()) ? ${paramName} : System.getenv("${paramName.toUpperCase()}");`;
    const ifQueryProvided = `if (this.${paramName} != null){`;
    const assignment = `params.put("${paramName}", this.${paramName});`;
    const closingTag = '}';
    return (
      <>
        <Text indent={6}>
          {variableDefinition}
        </Text>
        <Text indent={6}>
          {ifQueryProvided}
        </Text>
        <Text indent={8}>
          {assignment}
        </Text>
        <Text indent={6} newLines={1}>
          {closingTag}
        </Text>
      </>
    );
  });
}
