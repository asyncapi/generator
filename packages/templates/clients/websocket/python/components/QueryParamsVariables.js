import { Text } from '@asyncapi/generator-react-sdk';

export function QueryParamsVariables({ queryParams }) {
  if (!queryParams) {
    return null;
  }

  return queryParams.map((param) => {
    const paramName = param[0];
    const variableDefinition = `${paramName} = ${paramName} or os.getenv("${paramName.toUpperCase()}")`;
    const ifQueryProvided = `if ${paramName} is not None:`;
    const assignment = `params["${paramName}"] = ${paramName}`;
    return (
      <>
        <Text indent={8}>
          {variableDefinition}
        </Text>
        <Text indent={8}>
          {ifQueryProvided}
        </Text>
        <Text indent={10}>
          {assignment}
        </Text>
      </>
    );
  });
}
