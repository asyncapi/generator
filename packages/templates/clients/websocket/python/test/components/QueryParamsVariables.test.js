import path from 'path';
import { render } from '@asyncapi/generator-react-sdk';
import { Parser, fromFile } from '@asyncapi/parser';
import { getQueryParams } from '@asyncapi/generator-helpers';
import { QueryParamsVariables } from '../../components/QueryParamsVariables.js';

const parser = new Parser();
const asyncapiFilePath = path.resolve(
  __dirname,
  '../../../../../../helpers/test/__fixtures__/asyncapi-websocket-query.yml'
);

describe('QueryParamsVariables component (integration with AsyncAPI document)', () => {
  let parsedAsyncAPIDocument;

  beforeAll(async () => {
    try {
      const parseResult = await fromFile(parser, asyncapiFilePath).parse();
      parsedAsyncAPIDocument = parseResult.document;
    } catch (error) {
      throw new Error(`Failed to parse AsyncAPI document: ${error.message}`);
    }
  });

  test('renders correctly with query parameters', () => {
    const channels = parsedAsyncAPIDocument.channels();
    const queryParamsObject = getQueryParams(channels);
    const queryParamsArray = queryParamsObject ? Array.from(queryParamsObject.entries()) : [];
    const result = render(<QueryParamsVariables queryParams={queryParamsArray} />);
    expect(result.trim()).toMatchSnapshot();
  });

  test('renders nothing when queryParams is null', () => {
    const result = render(<QueryParamsVariables queryParams={null} />);
    expect(result).toBe('');
  });

  test('renders nothing when queryParams is undefined', () => {
    const result = render(<QueryParamsVariables />);
    expect(result).toBe('');
  });
});
