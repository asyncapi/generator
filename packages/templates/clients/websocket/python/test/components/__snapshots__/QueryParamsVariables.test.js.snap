// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`QueryParamsVariables component (integration with AsyncAPI document) renders correctly with query parameters 1`] = `
"heartbeat = heartbeat or os.getenv("HEARTBEAT")
        if heartbeat is not None:
          params["heartbeat"] = heartbeat
        top_of_book = top_of_book or os.getenv("TOP_OF_BOOK")
        if top_of_book is not None:
          params["top_of_book"] = top_of_book
        bids = bids or os.getenv("BIDS")
        if bids is not None:
          params["bids"] = bids
        offers = offers or os.getenv("OFFERS")
        if offers is not None:
          params["offers"] = offers"
`;
